sled-agent/src/services.rs

```diff
@@ -1194,6 +1197,155 @@ impl ServiceManager {
                 let running_zone = RunningZone::boot(installed_zone).await?;
                 return Ok(running_zone);
             }
+            ZoneType::Nexus => {
+                let (internal_ip, external_tls) = request
+                    .zone
+                    .services
+                    .iter()
+                    .find_map(|s| {
+                        if let ServiceType::Nexus {
+                            internal_ip,
+                            external_tls,
+                            ..
+                        } = s.details
+                        {
+                            Some((internal_ip, external_tls))
+                        } else {
+                            None
+                        }
+                    })
+                    .ok_or_else(|| {
+                        Error::BadRequest(
+                            "Nexus zone missing Nexus service".to_string(),
+                        )
+                    })?;
+
+                let sled_info = self
+                    .inner
+                    .sled_info
+                    .get()
+                    .ok_or(Error::SledAgentNotReady)?;
+
+                // While Nexus will be reachable via `external_ip`, it communicates
+                // atop an OPTE port which operates on a VPC private IP. OPTE will
+                // map the private IP to the external IP automatically.
+                let port =
+                    installed_zone.opte_ports().nth(0).ok_or_else(|| {
+                        Error::BadRequest(
+                            "Nexus zone missing OPTE port".to_string(),
+                        )
+                    })?;
+                let opte_interface = port.vnic_name();
+                let opte_gateway = &port.gateway().ip().to_string();
+
+                // Nexus takes a separate config file for parameters which
+                // cannot be known at packaging time.
+                let nexus_port = if external_tls { 443 } else { 80 };
+                let deployment_config = NexusDeploymentConfig {
+                    id: request.zone.id,
+                    rack_id: sled_info.rack_id,
+
+                    dropshot_external: ConfigDropshotWithTls {
+                        tls: external_tls,
+                        dropshot: dropshot::ConfigDropshot {
+                            // This is a bit silly, but:
+                            // - We do not know the private IP address of Nexus
+                            // until the zone is launched.
+                            // - To remediate, we pass a placeholder value here.
+                            // - We replace that value in the Nexus method
+                            // script. See the "external_ip_override" argument
+                            // to the Nexus binary.
+                            bind_address: SocketAddr::new(
```

Comment:
I don't love this, and it's probably worth a critical look.

Basically:
- Previously, we waited until the zone was running to create this IP address
- We did so with `ipadm create-addr -t -T dhcp "$OPTE_INTERFACE/public"`
- This created a new address which we plugged into the zone config

However, now that we're trying to plumb this config into the zone **before** it launches, the IP address doesn't exist, and we don't know what it is.

Reply:
We do know the private, you might have to plumb it up a little but the `Port` type has it: https://github.com/oxidecomputer/omicron/blob/15e441f9ac24f4b3d8221eca07e25fe261297b1a/illumos-utils/src/opte/port.rs#L14-L18

The "dhcp" server is just gonna return the ip configured at port creation. We could've easily just did the `ipadm` calls for a static ip but dhcp also plumbs the correct routes (except we add those manually because illumos' dhcp client doesn't support the classes static routes option).